pkg/cmd/cli/cmd/loginoptions.go

```diff
@@ -0,0 +1,325 @@
+
+	// 		if ctx.AuthInfo == o.Username && clusterInfo.Server == o.Server && len(authInfo.Token) > 0 { // only token for now
+	// 			glog.V(5).Infof("Authentication exists for '%v' on '%v', trying to use it...\n", o.Server, o.Username)
+
+	// 			o.Config.BearerToken = authInfo.Token
+
+	// 			if me, err := whoami(o.Config); err == nil && usernameFromUser(me) == o.Username {
+	// 				o.Username = usernameFromUser(me)
+	// 				return nil
+	// 			}
+
+	// 			glog.V(5).Infof("Token %v no longer valid for '%v', can't use it\n", authInfo.Token, o.Username)
+	// 		}
+	// 	}
+	// }
+
+	o.gatherAuthInfo = true
+	return nil
+}
+
+// Discover the projects available for the stabilished session and take one to use. It
+// fails in case of no existing projects, and print out useful information in case of
+// multiple projects.
+func (o *LoginOptions) GatherProjectInfo() error {
+	o.assertGatheredAuthInfo()
+
+	oClient, err := client.New(o.Config)
+	if err != nil {
+		return err
+	}
+
+	projects, err := oClient.Projects().List(labels.Everything(), labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	projectsItems := projects.Items
+
+	switch len(projectsItems) {
+	case 0:
+		if me, err := whoami(o.Config); err == nil {
+			return fmt.Errorf(`You don't have any project.
+To create a new project, run 'openshift ex new-project <projectname> --admin=%s'.
+To be added as an admin to an existing project, run 'openshift ex policy add-user admin %s -n <projectname>'.`, me.Name, me.Name)
+		}
+		return err
+
+	case 1:
+		o.Project = projectsItems[0].Name
+
+	default:
+		projects := []string{}
+
+		for _, project := range projectsItems {
+			projects = append(projects, project.Name)
+		}
+
+		namespace, err := o.ClientConfig.Namespace()
+		if err != nil {
+			return err
+		}
+
+		if current, err := oClient.Projects().Get(namespace); err != nil {
+			if kerrors.IsNotFound(err) || kerrors.IsForbidden(err) {
+				o.Project = projects[0]
+			}
+			return err
+		} else {
+			o.Project = current.Name
+		}
+
+		if n := len(projects); n > 10 {
+			projects = projects[:10]
+			fmt.Printf("You have %d projects, displaying only the first 10. To view all your projects run 'osc get projects'.\n", n)
+		}
+		var sortedProjects sort.StringSlice = projects
+		sortedProjects.Sort()
+		fmt.Printf("Your projects are: %v. You can switch between them at any time using 'osc project <project-name>'.\n", strings.Join(projects, ", "))
+	}
+
+	fmt.Printf("Using project '%v'.\n", o.Project)
+
+	o.gatherProjectInfo = true
+	return nil
+}
+
+// Save all the information present in this helper to a config file. An explicit config
+// file path can be provided, if not use the established conventions about config
+// loading rules. Will create a new config file if one can't be found at all. Will only
+// succeed if all required info is present.
+func (o *LoginOptions) SaveConfig(explicitFilePath string) (bool, error) {
+	if len(o.Project) == 0 || len(o.Username) == 0 {
```

Comment:
And `SaveConfig`, on the other hand, just checks for the presence of what it needs to save.